[RDY] Allow nil in list when decoding lsp responses

akinsho opened the pull request:
This PR is based on the discussion in #13829, where `nil` is being dropped whilst decoding the lsp server responses (and requests?). It simply uses `vim.tbl_islist` as suggested there. One question I still have is if we have to check that the list doesn't contain any tables that might need to have nil values converted. I'm not very familiar with the lsp implementation here so not sure if that's necessary.

It does however resolve the issue in #14564 and so will hopefully also fix #13829

@mjlbach was this what you had in mind regarding a fix for #13829

#### Todo
- [x] Check that this hasn't broken anything...

akinsho committed:
fix(lsp): do not convert vim.NIL to nil in lists

mjlbach commented:
Would be good to have a test :)

akinsho commented:
@mjlbach any chance you can point me towards where this test should go? I've read through the testing part of the README, still not entirely sure where this file is being tested, there's no `rpc_spec` or anything

mjlbach commented:
prob just lsp_spec.lua

akinsho commented on test/functional/plugin/lsp_spec.lua:
```diff
@@ -926,6 +926,49 @@ describe('LSP', function()
       }
     end)
+
+    it('should handle return nil when decoding a list', function()
```
@mjlbach sorry to be a bother but in the interest of time/trying to get this in for 0.5 I wonder if you could help. The testing helpers for the lsp presume a much better understanding of neovim's client than I have and I'd like to avoid losing too much time trying to wrangling with it.

I'm essentially just trying to send a request back and forth from the fake server and assert that `vim.NIL` is kept if the params contains a list but my test keeps failing because the expected callbacks are in the wrong order 🤔 . There's kind of a lot going on across the various test helpers

Is this the kind of test you had in mind, should it be something simpler that just checks the behaviour of `handle_body`?

mjlbach replied:
I can help debug any tests this weekend. I would try to match the callbacks expected by other tests, there shouldn't be anything unique about this other than the fact that you're using code-actions which has 2 round trips

akinsho replied:
@mjlbach I've spent more time plugging away at this now and I'm not really sure how specifically to test this based on the examples in `lsp_spec.lua`. It doesn't have any docs so it's mostly been trial and error/reading through some of the helper implementations.

What should the flow be within my test case be? I tried basing it on the issue I saw i.e. I've tried triggering and responding to a code action which should trigger an `executeCommand` if this command contained a list with a `nil` value then if this change was working we'd expect it to get through to the test in `fake_lsp_server.lua` with the `vim.NIL` intact?

Trying to trigger the `vim.lsp.buf.code_action()` doesn't seem to do anything i.e. the `textDocument/codeAction` isn't sent to the fake lsp server and none of the subsequent commands are executed, I also can't tell whether it getting through would correctly validate that `NIL` was no longer being filtered out. Just passing notifications through using `client.notify` doesn't seems to make a difference whether I comment out my changes or not.

akinsho commented:
@mjlbach apologies for all the pings. I *think* I finally got a handle on the test setup and have updated my test although it doesn't really seem to matter whether or not my change is included or not the tests seem to pass regardless i.e. I send a request from the client with a list containing `vim.NIL` and whether or not I have my changes in or not the test passes.

The tests all seem to rebuild nvim every time so I'm guessing nothing is cached and looking at `client.request` it seems as though it would flow through the code path I changed but it doesn't seem to make a difference in either case. Any ideas?

mfussenegger commented on runtime/lua/vim/lsp/rpc.lua:
```diff
@@ -50,9 +50,11 @@ recursive_convert_NIL = function(v, tbl_processed)
-    return vim.tbl_map(function(x)
-      return recursive_convert_NIL(x, tbl_processed)
-    end, v)
+    if not vim.tbl_islist(v) then
```
This will also mean that dictionaries within tables will no longer be recursively processed. Maybe we should acount for that and only keep `vim.NIL` in lists, but still recurse into dictionary-style-tables ?

akinsho replied:
Just pushed a changed which should allow trimming `vim.NIL` in dictionaries but spare lists 👍🏾

mjlbach commented:
@akinsho the PR LGTM but the test included seems to not fail even without this PR. The fix should check that a table terminating with `nil` has the correct number of arguments when sent from the server to the client.

akinsho commented:
@mjlbach I've tweaked the test and verified that it now fails without this PR's change. I think this is done on my end. Let me know if there's anything missing.

mjlbach commented:
@mfussenegger this LGTM now, WDYT?

@akinsho thanks for driving this. Can you squash your commits and use conventional commit style on the message? Otherwise I'll squash/merge when tests are green

akinsho commented:
@mjlbach I've already squashed them down and am using the conventional commit syntax

mjlbach commented:
I thrink a more accurate commit message is `fix(lsp): do not convert vim.NIL to nil in lists`

akinsho commented:
I've rebased the branch off master and updated the test. I think the reference to callbacks in the lsp spec changed at some point and I needed to update this PR. Not sure if that was why it was failing, but it seems to pass all functional tests on my machine now
